File: day11/main.py
from argparse import ArgumentParser
from collections import Counter
from pathlib import Path


def is_valid_point(x, y, data):
    return x >= 0 and x < len(data) and y >= 0 and y < len(data[x])


def neighbors(i, j, data):
    for x in range(i-1, i+2):
        for y in range(j-1, j+2):
            if (x, y) != (i, j) and is_valid_point(x, y, data):
                yield x, y


def do_flashing(i, j, data, flashed):

    def flash(i, j):
        if data[i][j] > 9 and (i, j) not in flashed:
            flashed.add((i, j))
            for x, y in neighbors(i, j, data):
                data[x][y] += 1
                flash(x, y)

    flash(i, j)


if __name__ == "__main__":

    parser = ArgumentParser()
    parser.add_argument("-i", "--infile", required=False, type=Path, default=Path("input"))
    args = parser.parse_args()

    lines = Path(args.infile).read_text().rstrip("\n").split("\n")

    data = [[int(x) for x in row] for row in lines]
    size = len(data)*len(data[0])

    flashes = Counter()
    step = 1
    flashed = set()

    while len(list(flashed)) != size:

        data = [[x + 1 for x in row] for row in data]

        flashed = set()
        for i, row in enumerate(data):
            for j, _ in enumerate(row):
                do_flashing(i, j, data, flashed)

        flashes[step] += flashes[step-1]
        for x, y in list(flashed):
            data[x][y] = 0
            flashes[step] += 1

        step += 1

    print(f"Part 1 result: {flashes[100]}")
    print(f"Part 2 result: {step-1}")

File: day11/main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"strings"
)

type pos struct {
	x int
	y int
}

func FlashSequence(i int, j int, data [][]int, flashed map[pos]struct{}) {

	var flash func(i int, j int)
	flash = func(i int, j int) {
		if data[i][j] > 9 {
			if _, ok := flashed[pos{i, j}]; !ok {
				flashed[pos{i, j}] = struct{}{}
				for x := i - 1; x < i+2; x++ {
					for y := j - 1; y < j+2; y++ {
						if (pos{x, y} != pos{i, j}) && x >= 0 && x < len(data) && y >= 0 && y < len(data[x]) {
							data[x][y] += 1
							flash(x, y)
						}
					}
				}
			}
		}
	}

	flash(i, j)
}

func ReadMatrixInts(file string) [][]int {

	content, err := os.ReadFile(file)
	if err != nil {
		log.Fatal(err)
	}

	lines := strings.Split(strings.Trim(string(content), "\n"), "\n")
	data := make([][]int, len(lines))

	for i, x := range lines {
		data[i] = make([]int, len(x))
		for j, y := range x {
			data[i][j] = int(y) - '0'
		}

	}

	return data
}

func IncrementAll(data [][]int) [][]int {

	for i, row := range data {
		for j := range row {
			data[i][j] += 1
		}
	}

	return data
}

func main() {
	var file string
	flag.StringVar(&file, "infile", "input", "Input file")
	flag.Parse()

	data := ReadMatrixInts(file)

	flashes := 0
	for step := 0; step < 100; step++ {

		flashed := make(map[pos]struct{})
		data = IncrementAll(data)

		for i, row := range data {
			for j := range row {
				FlashSequence(i, j, data, flashed)
			}
		}

		for p := range flashed {
			data[p.x][p.y] = 0
			flashes += 1
		}
	}

	fmt.Println(flashes)
}
